flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/history/HistoryServerTest.java

```diff
@@ -117,45 +124,145 @@ public void testHistoryServerIntegration() throws Exception {
+		int numExpiredJobs = cleanupExpiredJobs ? 1 : 0;
+		int numJobs = 3;
+		List<String> jobIDs = new ArrayList<>();
+		for (int x = 0; x < numJobs; x++) {
+			jobIDs.add(runJob());
 		}
+		waitForArchivesCreation(numJobs);
+
+		CountDownLatch numExpectedArchivedJobs = new CountDownLatch(numJobs);
+		CountDownLatch numExpectedExpiredJobs = new CountDownLatch(numExpiredJobs);
+
+		Configuration historyServerConfig = createTestConfiguration(cleanupExpiredJobs);
+
+		HistoryServer hs =
+			new HistoryServer(
+				historyServerConfig,
+				(event) -> {
+					switch (event.getType()){
+						case CREATED:
+							numExpectedArchivedJobs.countDown();
+							break;
+						case DELETED:
+							numExpectedExpiredJobs.countDown();
+							break;
+					}
+				});
 
-		HistoryServer hs = new HistoryServer(historyServerConfig, numExpectedArchivedJobs);
 		try {
 			hs.start();
+			// check that expected count of archives exists
 			String baseUrl = "http://localhost:" + hs.getWebPort();
 			numExpectedArchivedJobs.await(10L, TimeUnit.SECONDS);
 
 			ObjectMapper mapper = new ObjectMapper();
-			String response = getFromHTTP(baseUrl + JobsOverviewHeaders.URL);
-			MultipleJobsDetails overview = mapper.readValue(response, MultipleJobsDetails.class);
+			Assert.assertEquals(numJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+
+			String jobIDToDelete = jobIDs.get(0);
+			verifyArchiveFilesExistence(jobIDToDelete, true);
+
+			// delete one archive from jm
+			URI resolve = jmDirectory.toURI().resolve(jobIDToDelete);
+			new File(resolve).delete();
+
+			numExpectedExpiredJobs.await(10L, TimeUnit.SECONDS);
+
+			// check that archive presence in hs
+			Assert.assertEquals(numJobs - numExpiredJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+			verifyArchiveFilesExistence(jobIDToDelete, !cleanupExpiredJobs);
 
-			Assert.assertEquals(numJobs + 1, overview.getJobs().size());
 		} finally {
 			hs.stop();
 		}
 	}
 
-	private static void runJob() throws Exception {
+	private void verifyArchiveFilesExistence(String jobID, boolean expectArchiveFilesExist) {
+		List<File> archiveFiles = Arrays.asList(
+			new File(hsDirectory.toURI().resolve(String.format("overviews/%s.json", jobID))),
+			new File(hsDirectory.toURI().resolve(String.format("jobs/%s", jobID))),
+			new File(hsDirectory.toURI().resolve(String.format("jobs/%s.json", jobID))));
+
+		archiveFiles.forEach(
+			archiveFile -> Assert.assertEquals(
+				String.format("Existence of file or directory %s", archiveFile.getPath()),
+				archiveFile.exists(), expectArchiveFilesExist));
+	}
+
+	private void waitForArchivesCreation(int numJobs) throws InterruptedException {
+		// the job is archived asynchronously after env.execute() returns
+		File[] archives = jmDirectory.listFiles();
+		while (archives == null || archives.length != numJobs) {
+			Thread.sleep(50);
+			archives = jmDirectory.listFiles();
+		}
+	}
+
+	private Configuration createTestConfiguration(boolean cleanupExpiredJobs) {
```

Comment:
I would prefer these refactorings to be moved into a separate commit, including
* `waitForArchivesCreation`
* `createTestConfiguration`
* `runJob`
* `createLegacyArchive`

Reply:
pr is separated into two commits. Thank you @aljoscha for helping with that